core/src/calleeFunctions/index.ts

```diff
@@ -23,27 +23,101 @@ const allCalleeFunctions: Record<CalleeNames, CalleeFunctions> = {
 export const getCalleeData = async function (
     network: string,
     collateralType: string,
+    marketId: string,
     profitAddress: string
 ): Promise<string> {
     const collateral = getCollateralConfigByType(collateralType);
-    const calleeFuctions = allCalleeFunctions[collateral.exchange.callee];
-    if (!calleeFuctions) {
+    const marketData = collateral.exchanges[marketId];
+    if (!marketData || !marketData.callee || !allCalleeFunctions[marketData.callee]) {
         throw new Error(`Unsupported collateral type "${collateralType}"`);
     }
-    return await calleeFuctions.getCalleeData(network, collateral, profitAddress);
+    return await allCalleeFunctions[marketData.callee].getCalleeData(network, collateral, marketId, profitAddress);
 };
 
-const _getMarketPrice = async function (
+export const getMarketData = async function (
     network: string,
     collateralSymbol: string,
     amount: BigNumber = new BigNumber('1')
-): Promise<BigNumber> {
+): Promise<Record<string, MarketData>> {
     const collateral = getCollateralConfigBySymbol(collateralSymbol);
-    const calleeFuctions = allCalleeFunctions[collateral.exchange.callee];
-    if (!calleeFuctions) {
+    const isCollateralSupported = Object.values(collateral.exchanges).some(
+        marketData => marketData.callee && allCalleeFunctions[marketData.callee]
+    );
+    if (!isCollateralSupported) {
         throw new Error(`Unsupported collateral symbol "${collateralSymbol}"`);
     }
-    return await calleeFuctions.getMarketPrice(network, collateral, amount);
+    let marketDataRecords = {};
+    for (const marketId in collateral.exchanges) {
+        const marketData = collateral.exchanges[marketId];
+        let marketUnitPrice: BigNumber;
+        try {
+            marketUnitPrice = await allCalleeFunctions[marketData.callee].getMarketPrice(
+                network,
+                collateral,
+                marketId,
+                amount
+            );
+        } catch {
+            marketUnitPrice = new BigNumber(NaN);
+        }
```

Comment:
Why do you need to try/catch it here? (if your function can still `throw`)

In other words, if you want to return market data records in _any_ case, you need to move `Unsupported collateral symbol "${collateralSymbol}"` to be thrown on the market level?

I also suggest to add new optional key `marketData.errorMessage` which will contain those errors (because previously it was cached on the frontend level and we were able to display this error)

Reply:
I still don't see any `errorMessage` inside `marketData`